### DISCUSSION TITLE
cwebp optimizer supports only webp images?

### DISCUSSION THREAD

**Discussion by tomaszkane** [spatie/image-optimizer#192] [Ideas]
Why Cwebp optimizer accept only webp?
https://github.com/spatie/image-optimizer/blob/main/src/Optimizers/Cwebp.php#L13

`cwebp` binary supports jpg, png, tiff and webp as input file
https://developers.google.com/speed/webp/docs/cwebp?hl=en#description

**Comment by tomaszkane** [comment]
IMHO it should looks like:
```php
public function canHandle(Image $image): bool
{
    return in_array($image->mime(), ['image/webp', 'image/jpeg', 'image/png', /* ...... */]);
}
```

**Reply by redburn** [reply]
This package cannot convert between image formats, unfortunately. It always loops over _all_ the optimizers' canHandle() methods and then (in a chain) runs any optimizer that says it can handle the file's mime type. As such, your version of canHandle() would result in cwebp optimization (and conversion!) being applied to any image/png file you pass to optimize(). Probably not what you're after.

Given that most WEBP and AVIF files are still conversions of existing PNG and JPEG files, I agree that it would be a helpful feature.

I worked around this by writing a custom optimizer and optimization chain. It's essentially a copy of the Cwebp optimizer, but I've changed the canHandle() method to check for the .webp extension and for a valid mime type.

```
use Spatie\ImageOptimizer\Image;
use Spatie\ImageOptimizer\Optimizers\Cwebp;

class ConvertToWebp extends Cwebp {

    private array $supportedMimeTypes = ['image/jpeg', 'image/png', 'image/tiff', 'image/webp'];

    public function canHandle(Image $image): bool
    {
        return ($image->extension() === 'webp' && in_array($image->mime(), $this->supportedMimeTypes));
    }

}
```
Then simply pass the above optimizer to a fresh OptimizerChain:
```
$webpConverter = (new OptimizerChain)
    ->addOptimizer(
        new ConvertToWebp()
    );

$webpConverter->optimize('image.jpg', 'image.jpg.webp');
```
This creates an optimized WEBP version of the source image in one swoop.
